pytket/extensions/cutensornet/mps/mps.py

```diff
@@ -0,0 +1,735 @@
+try:
+    import cuquantum as cq  # type: ignore
+    import cuquantum.cutensornet as cutn  # type: ignore
+except ImportError:
+    warnings.warn("local settings failed to import cutensornet", ImportWarning)
+
+from pytket.circuit import Command, Op, Qubit  # type: ignore
+
+# An alias so that `intptr_t` from CuQuantum's API (which is not available in
+# base python) has some meaningful type name.
+Handle = int
+# An alias for the type of the unique identifiers of each of the bonds
+# of the MPS.
+Bond = int
+
+
+class DirectionMPS(Enum):
+    """An enum to refer to relative directions within the MPS."""
+
+    LEFT = 0
+    RIGHT = 1
+
+
+class Tensor:
+    """Class for the management of tensors via CuPy and cuTensorNet.
+    It abstracts away some of the low-level API of cuTensorNet.
+
+    Attributes:
+        data (cupy.ndarray): The entries of the tensor arranged in a CuPy ndarray.
+        bonds (list[Bond]): A list of IDs for each bond, matching the same order
+            as in ``self.data.shape`` (which provides the dimension of each).
+        canonical_form (Optional[DirectionMPS]): If in canonical form, indicate
+            which one; else None.
+    """
+
+    def __init__(self, data: cp.ndarray, bonds: list[Bond]):
+        """
+        Args:
+            data: The entries of the tensor arranged in a CuPy ndarray.
+            bonds: A list of IDs for each bond, matching the same order
+                as in ``self.data.shape`` (which provides the dimension of each).
+        """
+        self.data = data
+        self.bonds = bonds
+        self.canonical_form: Optional[DirectionMPS] = None
+
+    def get_tensor_descriptor(self, libhandle: Any) -> Handle:
+        """Return the cuTensorNet tensor descriptor.
+
+        Note:
+            The user is responsible of destroying the descriptor once
+            not in use (see ``cuquantum.cutensornet.destroy_tensor_descriptor``).
+
+        Args:
+            libhandle: The cuTensorNet library handle.
+
+        Returns:
+            The handle to the tensor descriptor.
+        """
+        if libhandle is None:
+            raise RuntimeError(
+                "Must be called inside a with mps.init_cutensornet() block."
+            )
+        if self.data.dtype == np.float32:
+            cq_dtype = cq.cudaDataType.CUDA_R_32F
+        elif self.data.dtype == np.float64:
+            cq_dtype = cq.cudaDataType.CUDA_R_64F
+        elif self.data.dtype == np.complex64:
+            cq_dtype = cq.cudaDataType.CUDA_C_32F
+        elif self.data.dtype == np.complex128:
+            cq_dtype = cq.cudaDataType.CUDA_C_64F
+
+        return cutn.create_tensor_descriptor(  # type: ignore
+            handle=libhandle,
+            n_modes=len(self.data.shape),
+            extents=self.data.shape,
+            strides=self.get_cuquantum_strides(),
+            modes=self.bonds,
+            data_type=cq_dtype,
+        )
+
+    def get_cupy_strides(self) -> list[int]:

```

Comment:
Since it hasn't been used yet, I doubt it'll be needed -- I added it at the very beginning of coding this up, thinking I'd need it. I'm happy to remove it.

Yes, I'm happy to go over the functions that I don't want in the public API and add an underscore to their name.